File: internal/model/poisson.go
package model

import (
	"math"
)

type poissonCacheKey struct {
	lambda float64 
	k      int
}

func newPoissonCache(lambda float64, k int) poissonCacheKey {
	return poissonCacheKey{lambda: lambda, k: k}
}

type PoissonModel struct {
	MaxGoals      int
	HomeAdvantage float64
	Cache          map[poissonCacheKey]float64	
}

func NewPoissonModel(MaxGoals int, HomeAdvantage float64) *PoissonModel {
	return &PoissonModel{
		MaxGoals:      MaxGoals,
		HomeAdvantage: HomeAdvantage,
		Cache: make(map[poissonCacheKey]float64),
	}
}

func factorial(n int) int {
	result := 1
	for i := 2; i <= n; i++ {
	    result *= i
	}
	return result
  }

func (pm *PoissonModel) CalcProb(lambda float64, k int) float64 {
	key := newPoissonCache(lambda, k)
	if value, exists := pm.Cache[key]; exists{
		return value
	}

	p := (math.Pow(lambda, float64(k)) * math.Exp(-lambda)) / float64(factorial(k))
	pm.Cache[key] = p

	return p
}

func (pm *PoissonModel) GoalDistribution(lambda float64) []float64{
	probs := make([]float64, pm.MaxGoals+1)
	for k := 0; k <= pm.MaxGoals; k++ {
		probs[k] = pm.CalcProb(lambda, k)
	}

	total := 0.0
	for _, p := range probs {
		total += p
	}

	normalized := make([]float64, len(probs))
	for i, p := range probs {
		normalized[i] = p / total
	}

	return normalized
}

func (pm *PoissonModel) ScoreMatrix(lambdaHome, lambdaAway float64) [][] float64{
	homeDist := pm.GoalDistribution(lambdaHome * pm.HomeAdvantage)
	awayDist := pm.GoalDistribution(lambdaAway)

	matrix := make([][]float64, pm.MaxGoals + 1)
	for i:= 0; i <= pm.MaxGoals; i++ {
		matrix[i] = make([]float64, pm.MaxGoals + 1)

		for j := 0; j <= pm.MaxGoals; j++ {
			matrix[i][j] = homeDist[i] * awayDist[j]
		}
	}

	return matrix
}

func (pm *PoissonModel) MatchOutcome([][]float64) map[string]float64 {
	outcome := map[string]float64 {
		"home":0.0,
		"draw":0.0,
		"away":0.0,
	}

	for i := 0; i <= pm.MaxGoals; i++ {
		for j := 0; j <= pm.MaxGoals; j++ {
			if i > j {
				outcome["home"] += matrix[i][j]
			} else if i == j {
				outcome["draw"] += matrix[i][j]
			} else {
				outcome["away"] += matrix[i][j]
			}
		}
	}

	return outcome
}

File: core/model/poisson.py
import math
from dataclasses import dataclass, field

@dataclass
class PoissonModel:
    max_goals: int
    cache: dict[tuple[float, int], float] = field(default_factory=dict)

    def calcProb(self, lmbda: float, k: int) -> float:
        key = (round(lmbda, 4), k)
        if key in self.cache:
            return self.cache[key]

        log_p = k * math.log(lmbda) - lmbda - math.lgamma(k + 1)
        p = math.exp(log_p)
        self.cache[key] = p
        return p

    def goalDistribution(self, lmbda: float) -> list[float]:
        probs = [self.calcProb(lmbda, k) for k in range(self.max_goals + 1)]
        total = sum(probs)
        return [p / total for p in probs]
    
    def scoreMatrix(self, lambda_home: float, lambda_away: float) -> list[list[float]]:
        home_probs = self.goalDistribution(lambda_home)
        away_probs = self.goalDistribution(lambda_away)

        matrix = [[home_probs[i] * away_probs[j] for j in range(self.max_goals + 1)] for i in range(self.max_goals + 1)]
        return matrix
    
    def matchOutcome(self, lambda_home: float, lambda_away: float) -> dict[str, float]:
        matrix = self.scoreMatrix(lambda_home, lambda_away)
        outcome = {
            "home": 0.0,
            "draw": 0.0,
            "away": 0.0,
        }

        for i in range(self.max_goals + 1):
            for j in range(self.max_goals + 1):
                if i > j:
                    outcome["home"] += matrix[i][j]
                elif i == j:
                    outcome["draw"] += matrix[i][j]
                else:
                    outcome["away"] += matrix[i][j]
        return outcome
    
    def probToOdd(self, p: float) -> float:
        return 1.0 / p
    
    def getMatchOdds(self, lambda_home: float, lambda_away: float) -> dict[str, float]:
        probs = self.matchOutcome(lambda_home, lambda_away)
        
        odds = {
            label: self.probToOdd(p) if p > 0 else float('inf') 
            for label, p in probs.items()
        }
        
        return odds